### PR TITLE
Ignore property that does not exist for interfaces (identifiers for interfaces)

### PR DESCRIPTION
**Pull request by Nek-**
| Q             | A
| ------------- | ---
| Bug fix?      | yes/no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | Related to https://github.com/api-platform/core/issues/1574
| License       | MIT
| Doc PR        | N/A

### Motivations

TL;DR: Before: no "item" route working because we can't define an identifier. After: we can define a property for an interface and it will not be considered as "not found property".

1. It makes no sense to check if an interface exists in case it's an interface ;
2. If the property is not recognized as one and there is no identifier, then API Platform will not be able to call the DataProvider for a given item and interface as a resource is an incomplete feature.

### Usage

```yaml
Me\InterfaceAsAResource:
    properties:
         # Will work even if foo is not defined in the interface!
         foo:
             identifier: true
```

### What's next?

If you like this kind of features, I will probably work on "virtual properties" next. I see something like

```yaml
Me\InterfaceAsAResource:
    properties:
         foo:
             identifier: true
             proxy_method: 'customMethodToGetFoo'
```

What do you think? Maybe you want a separate RFC issue to discuss it?

### CODE REVIEW

File: src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php [hunk 1]
```diff
@@ -48,9 +48,15 @@ public function create(string $resourceClass, string $property, array $options =
             }
         }
 
+        try {
+            $isInterface = (new \ReflectionClass($resourceClass))->isInterface();
+        } catch (\ReflectionException $e) {
+            $isInterface = false;
+        }
+
         if (
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: `!$isInterface` costs less, you can check it first.

Comment (Nek-) [reply src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: Fixed 👍

**Comment by teohhanhui** [comment]
I thought we already support virtual properties?

**Comment by teohhanhui** [comment]
@Nek- I think there's no documentation, not even tests for it. :laughing: :see_no_evil:

But looking at the code (and as I've heard from other's experience), it should work. (Not when using the `Extractor` factory, obviously, as you've discovered.)

**Comment by Nek-** [comment]
In any case, it doesn't invalidate this feature as I understand. Are failures on CI related? (due to inversion of condition?)

**Comment by teohhanhui** [comment]
CI failures are unrelated and should be fixed after https://github.com/api-platform/core/pull/2106 is merged.

File: tests/Fixtures/FileConfigurations/interface_resource.yml [hunk 2]
```diff
@@ -1,2 +1,8 @@
 resources:
-    'ApiPlatform\Core\Tests\Fixtures\DummyResourceInterface': ~
```
Comment (soyuka) [inline tests/Fixtures/FileConfigurations/interface_resource.yml]: shouldn't we keep a test like this?

**Comment by Nek-** [comment]
It's up to date and issues non-related.

File: src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php [hunk 3]
```diff
@@ -48,9 +48,15 @@ public function create(string $resourceClass, string $property, array $options =
             }
         }
 
+        try {
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: Can't this entire block be replaced by a simple `interface_exists($resourceClass)` call?

Comment (Nek-) [reply src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: Fixed

**Comment by soyuka** [comment]
I think that https://github.com/api-platform/core/pull/2495 fixes this case lmk if it doesn't.

Going to cherry pick the tests anyway!

**Comment by Nek-** [comment]
@soyuka it does not.

File: src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php [hunk 4]
```diff
@@ -48,9 +48,11 @@ public function create(string $resourceClass, string $property, array $options =
             }
         }
 
+        $isInterface = interface_exists($resourceClass);
+
         if (
-            !property_exists($resourceClass, $property) ||
-            !$propertyMetadata = $this->extractor->getResources()[$resourceClass]['properties'][$property] ?? false
+            !property_exists($resourceClass, $property) && !$isInterface ||
+            empty($propertyMetadata = $this->extractor->getResources()[$resourceClass]['properties'][$property] ?? [])
```
Comment (soyuka) [inline src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: null === (... ?? null)

Comment (Nek-) [reply src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: :+1: you right. (can't remember why I did it like that, probably a mistake after testing things)

**Pull request merged**